federation/apis/federation/types.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package federation
+
+import (
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+)
+
+// ClusterSpec describes the attributes on a Cluster.
+type ClusterSpec struct {
+	// a map of client CIDR to server address
+	ServerAddressByClientCIDRs []unversioned.ServerAddressByClientCIDR `json:"serverAddressByClientCIDRs"`
+	// the type (e.g. bearer token, client certificate etc) and data of the credential used to access cluster.
+	// It’s used for system routines (not behalf of users)
+	Credential string `json:"credential",omitempty`
+}
+
+type ClusterPhase string
+
+// These are the valid phases of a cluster.
+const (
+	// Newly registered clusters or clusters suspended by admin for various reasons. They are not eligible for accepting workloads
+	ClusterPending ClusterPhase = "pending"
+	// Clusters in normal status that can accept workloads
+	ClusterRunning ClusterPhase = "running"
+	// Clusters temporarily down or not reachable
+	ClusterOffline ClusterPhase = "offline"
+	// Clusters removed from federation
+	ClusterTerminated ClusterPhase = "terminated"
+)
+
+// Cluster metadata
+type ClusterMeta struct {
+	// Version of the cluster
+	Version string `json:"version,omitempty"`
+}
+
+// ClusterStatus is information about the current status of a cluster.
+type ClusterStatus struct {
+	// Phase is the recently observed lifecycle phase of the cluster.
+	Phase ClusterPhase `json:"phase,omitempty"`
```

Comment:
@bgrant0607 I must confess that I don't fully grok the argument against phase/state given in #7856.  I buy that adding new states to an existing state machine over time can break clients.  And that trying to dream up all possible states ahead of time can lead to unnecessary complexity.  But in this case the number of states is 4, and we don't envisage changing this, ever. This is the essence of @thockin 's argument in https://github.com/kubernetes/kubernetes/issues/7856#issuecomment-99657942 (I think).

I need to think about this some more, and/or have an in-person discussion with you on how best to proceed here.